### PR TITLE
extending the cordon command for all clusters

### PR DESCRIPTION
**Pull request by 09harsh**
# Description

*Please include a summary of the change and which issue is addressed. If this change resolves an issue, please include the issue number in the description.*

Currently the `cordon` command in plugin requires a cluster name to remove all the pods that are running of a specific node on a cluster. This PR is to extend that behaviour to remove all the pods that are running on that node irrespective of the cluster name. This requires us to use a new flag `use-custom-label` which enables this new feature and then we also need to provide a `custom-label`. The default value of `custom-label` is `fdb-cluster-name` is nothing was given. 

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

*Are there any design details that you would like to discuss further?*
No

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

Added unit test

## Documentation

*Did you update relevant documentation within this repository?*
Yes

*If this change is adding new functionality, do we need to describe it in our user manual?*

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

*Does this introduce new defaults that we should re-evaluate in the future?*

### CODE REVIEW

File: kubectl-fdb/cmd/cordon.go [hunk 1]
```diff
@@ -67,6 +66,14 @@ func newCordonCmd(streams genericclioptions.IOStreams) *cobra.Command {
 			if err != nil {
 				return err
 			}
+			useCustomLabel, err := cmd.Flags().GetBool("use-custom-label")
+			if err != nil {
+				return err
+			}
+			customLabel, err := cmd.Flags().GetString("custom-label")
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:73]:
```suggestion
			customLabel, err := cmd.Flags().GetString("cluster-label")
```

Why should we name it `custom-label` and not `cluster-label`?

File: kubectl-fdb/cmd/cordon.go [hunk 2]
```diff
@@ -67,6 +66,14 @@ func newCordonCmd(streams genericclioptions.IOStreams) *cobra.Command {
 			if err != nil {
 				return err
 			}
+			useCustomLabel, err := cmd.Flags().GetBool("use-custom-label")
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:69]: Is this additional flag really required?

File: kubectl-fdb/cmd/cordon.go [hunk 3]
```diff
@@ -116,18 +139,15 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
 	cmd.Flags().StringP("fdb-cluster", "c", "", "evacuate process group(s) from the provided cluster.")
 	cmd.Flags().StringToStringVarP(&nodeSelectors, "node-selector", "", nil, "node-selector to select all nodes that should be cordoned. Can't be used with specific nodes.")
 	cmd.Flags().BoolP("exclusion", "e", true, "define if the process groups should be removed with exclusion.")
-	err := cmd.MarkFlagRequired("fdb-cluster")
-	if err != nil {
-		log.Fatal(err)
-	}
-
+	cmd.Flags().Bool("use-custom-label", false, "define if the process groups should be removed using label instead of cluster name.")
+	cmd.Flags().StringP("custom-label", "l", "fdb-cluster-name", "evacuate process group(s) from the provided cluster.")
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:143]:
```suggestion
	cmd.Flags().StringP("custom-label", "l", fdbv1beta2. FDBClusterLabel, "evacuate process group(s) from the provided cluster.")
```

We should use the default label here if we define a default value for the flag. If we want to allow multiple custom labels we should make use of `StringArray` instead of a String and doing some custom parsing.

File: kubectl-fdb/cmd/cordon.go [hunk 4]
```diff
@@ -137,13 +157,22 @@ func cordonNode(kubeClient client.Client, cluster *fdbv1beta2.FoundationDBCluste
...
-			client.MatchingLabels(cluster.GetMatchLabels()),
-			client.MatchingFieldsSelector{
-				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
-			})
-
+		var err error
+		if useCustomLabel {
+			err = kubeClient.List(ctx.Background(), &pods,
+				client.InNamespace(namespace),
+				client.HasLabels(strings.Split(customLabel, " ")),
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:164]: I don't see how the cluster is dynamically fetched here.

**Review by johscheuer — CHANGES_REQUESTED** [review]
Have you tried running this command with the new changes? I believe it won't work. Currently it's still required to pass the cluster otherwise the command will fail. In addition we never dynamically fetch the cluster to add the process groups fro replacement. The current implementation would add all process groups to the one cluster that is selected with `-c`.

File: kubectl-fdb/cmd/cordon_test.go [hunk 5]
```diff
@@ -100,41 +101,80 @@ var _ = Describe("[plugin] cordon command", func() {
...
+					customLabel:    fdbv1beta2.FDBClusterLabel,
+				}),
+			Entry("Cordon no nodes with exclusion with custom label",
+				testCase{
+					nodes:                     []string{""},
+					WithExclusion:             true,
+					ExpectedInstancesToRemove: []string{},
+					ExpectedInstancesToRemoveWithoutExclusion: []string{},
+					useCustomLabel: true,
+					customLabel:    fdbv1beta2.FDBClusterLabel,
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:177]: Can we add some test cases where we remove multiple Pods from different clusters? The current test case is not enough to test the behaviour with multiple Pods from different clusters running on the same node.

**Review by johscheuer — CHANGES_REQUESTED** [review]

File: kubectl-fdb/cmd/cordon.go (hunk 1 above)
Comment (09harsh) [reply kubectl-fdb/cmd/cordon.go:66]: The idea is to allow other labels as well as per user requirement. 
Eg:
```
 kubectl fdb cordon -l "fdb-cluster-name some-other-label" node-1
 ```

File: kubectl-fdb/cmd/suite_test.go [hunk 6]
```diff
@@ -51,3 +41,17 @@ var _ = JustBeforeEach(func() {
 var _ = AfterEach(func() {
 	k8sClient.Clear()
 })
+
+func createCluster(givenName string, givenNamespace string) *fdbv1beta2.FoundationDBCluster {
```
Comment (johscheuer) [inline kubectl-fdb/cmd/suite_test.go:45]:
```suggestion
func createCluster(name string, namespace string) *fdbv1beta2.FoundationDBCluster {
```

Do you think the prefix "given" adds additional value to this parameters?

The name `createCluster` is a bit misleading as we are not really creating a cluster, we only generate the cluster struct (e.g. the `createPods` actually "creates" those Pods in the Kubernetes instance).

File: kubectl-fdb/cmd/cordon_test.go [hunk 7]
```diff
@@ -32,110 +32,190 @@ import (
...
+					ExpectedInstancesToRemove: []string{},
+					ExpectedInstancesToRemoveWithoutExclusion: []string{},
+					clusterName:  "",
+					clusterLabel: fdbv1beta2.FDBClusterLabel,
 				}),
 		)
 	})
 })
+
+func createPods(inputClusterName string, inputNamespace string, id int) error {
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:182]:
```suggestion
func createPods(clusterName string, namespace string, id int) error {
```

I think we can drop the prefix input here.

File: kubectl-fdb/cmd/cordon_test.go [hunk 8]
```diff
@@ -32,110 +32,190 @@ import (
...
+	pods := []corev1.Pod{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("instance-%d", id),
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:186]: Is there a reason we have to use the ID? It's probably better to prefix the Pod name with the cluster name, that way it's pretty obvious which Pods are associated with which cluster.

File: kubectl-fdb/cmd/cordon_test.go [hunk 9]
```diff
@@ -32,110 +32,190 @@ import (
...
+				err := cordonNode(k8sClient, input.clusterName, input.nodes, namespace, input.WithExclusion, false, 0, input.clusterLabel)
 				Expect(err).NotTo(HaveOccurred())
 
-				var resCluster fdbv1beta2.FoundationDBCluster
-				err = k8sClient.Get(context.Background(), client.ObjectKey{
-					Namespace: namespace,
-					Name:      clusterName,
-				}, &resCluster)
+				var clusterNames []string
+				if len(input.clusterName) == 0 {
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:65]:
```suggestion
				if input.clusterName == "" {
```

What is the reason for this additional check? Shouldn't we always check both clusters to ensure the code is not accidentally adding replacements to the second cluster even if it shouldn't?

File: kubectl-fdb/cmd/cordon_test.go [hunk 10]
```diff
@@ -32,110 +32,190 @@ import (
...
+					clusterNames = []string{clusterName, secondClusterName}
+				} else {
+					clusterNames = []string{input.clusterName}
+				}
+
+				var currentInstancesToRemove []string
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:71]:
```suggestion
				var instancesToRemove []string
```

File: kubectl-fdb/cmd/cordon_test.go [hunk 11]
```diff
@@ -32,110 +32,190 @@ import (
...
+				var currentInstancesToRemoveWithoutExclusion []string
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:72]:
```suggestion
				var instancesToRemoveWithoutExclusion []string
```

File: kubectl-fdb/cmd/cordon_test.go [hunk 12]
```diff
@@ -32,110 +32,190 @@ import (
...
+				for _, currentClusterName := range clusterNames {
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:73]:
```suggestion
				for _, clusterName := range clusterNames {
```

File: kubectl-fdb/cmd/cordon_test.go [hunk 13]
```diff
@@ -32,110 +32,190 @@ import (
...
+					clusterLabel: "",
 				}),
 			Entry("Cordon all nodes with exclusion",
 				testCase{
 					nodes:                     []string{"node-1", "node-2"},
 					WithExclusion:             true,
 					ExpectedInstancesToRemove: []string{"instance-1", "instance-2"},
 					ExpectedInstancesToRemoveWithoutExclusion: []string{},
+					clusterLabel: fdbv1beta2.FDBClusterLabel,
+					clusterName:  clusterName,
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:131]:
```suggestion
					clusterName:  clusterName,
					clusterLabel: fdbv1beta2.FDBClusterLabel,
```

**Review by johscheuer — COMMENTED** [review]
Have you tested those changes against a real cluster?

File: kubectl-fdb/cmd/cordon.go [hunk 14]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
+						cmd.Printf("could not fetch process group ID from Pod: %s\n", pod.Name)
+						continue
+					}
+					processGroups = append(processGroups, processGroup)
+				}
+			}
+			err = replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+			if err != nil {
+				operationFailed = true
+				errors = append(errors, fmt.Sprintf("unable to cordon all Pods for cluster %s\n", cluster.Name))
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:179]: Any reason not to print the error here directly?

File: kubectl-fdb/cmd/cordon.go [hunk 15]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
 			}
-			processGroups = append(processGroups, processGroup)
 		}
 	}
+	if operationFailed {
+		return fmt.Errorf("following operation failed, please check and retry: \n, %s", errors)
+	}
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:185]: If we already keep a list of errors, what is the benefit of having `operationFailed`? A check for `len(errors) > 0` should provide the same information.

File: kubectl-fdb/cmd/cordon.go [hunk 16]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
+	return nil
+}
 
-	return replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+func fetchPods(kubeClient client.Client, inputClusterName string, namespace string, node string, clusterLabel string) (corev1.PodList, error) {
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:189]:
```suggestion
func fetchPods(kubeClient client.Client, clusterName string, namespace string, node string, clusterLabel string) (corev1.PodList, error) {
```

File: kubectl-fdb/cmd/cordon.go [hunk 17]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
+		}
+		err = kubeClient.List(ctx.Background(), &pods,
+			client.InNamespace(namespace),
+			client.MatchingLabels(cluster.GetMatchLabels()),
+			client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
+			})
+	}
+	if err != nil {
+		return pods, fmt.Errorf("unable to fetch pods")
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:212]: Why are we not returning the actual error?

File: kubectl-fdb/cmd/cordon.go [hunk 18]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
+	var err error
+	if inputClusterName == "" {
+		err = kubeClient.List(ctx.Background(), &pods,
+			client.InNamespace(namespace),
+			client.HasLabels([]string{clusterLabel}),
+			client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
+			})
+	} else {
+		cluster, error := loadCluster(kubeClient, namespace, inputClusterName)
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:200]: Now we are fetching the cluster multiple times (once for every node). This is probably okay for now, but we should create an issue to resolve this.

File: kubectl-fdb/cmd/cordon.go [hunk 19]
```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
...
+func getClusterNames(cmd *cobra.Command, inputClusterName string, pods corev1.PodList, clusterLabel string) map[string]fdbv1beta2.None {
+	if inputClusterName != "" {
+		return map[string]fdbv1beta2.None{inputClusterName: {}}
+	}
+
+	clusterNames := make(map[string]fdbv1beta2.None)
+	for _, pod := range pods.Items {
+		clusterName, ok := pod.Labels[clusterLabel]
+		if !ok {
+			printStatement(cmd, fmt.Sprintf("could not fetch cluster name from Pod: %s\n", pod.Name), errorMessage)
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon.go:226]: I don't think we want to reuse this method here, this will always add a ❌ in front of the error message.

File: kubectl-fdb/cmd/cordon_test.go [hunk 20]
```diff
@@ -21,121 +21,219 @@
...
+				}
 
 				Expect(err).NotTo(HaveOccurred())
 				// Use equality.Semantic.DeepEqual here since the Equal check of gomega is to strict
-				Expect(equality.Semantic.DeepEqual(input.ExpectedInstancesToRemove, resCluster.Spec.ProcessGroupsToRemove)).To(BeTrue())
-				Expect(equality.Semantic.DeepEqual(input.ExpectedInstancesToRemoveWithoutExclusion, resCluster.Spec.ProcessGroupsToRemoveWithoutExclusion)).To(BeTrue())
+				fmt.Printf("\ninstancesToRemove: %v\n", instancesToRemove)
+				Expect(equality.Semantic.DeepEqual(input.ExpectedInstancesToRemove, instancesToRemove)).To(BeTrue())
+				fmt.Printf("\ninstancesToRemoveWithoutExclusion: %v\n", instancesToRemoveWithoutExclusion)
+				Expect(equality.Semantic.DeepEqual(input.ExpectedInstancesToRemoveWithoutExclusion, instancesToRemoveWithoutExclusion)).To(BeTrue())
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:90]:
```suggestion
				Expect(input.ExpectedInstancesToRemove).To(ConsistOf(instancesToRemove))
				Expect(input.ExpectedInstancesToRemoveWithoutExclusion).To(ConsistOf(instancesToRemoveWithoutExclusion))
```

See https://pkg.go.dev/github.com/onsi/gomega#ConsistOf, that check was "recently" added in gomega and provides more details on what element is missing.

File: kubectl-fdb/cmd/cordon_test.go [hunk 21]
```diff
@@ -21,121 +21,219 @@
...
+				testCase{
+					nodes:         []string{"node-1", "node-2"},
+					WithExclusion: true,
+					ExpectedInstancesToRemove: []string{
+						fmt.Sprintf("%s-instance-1", clusterName),
+						fmt.Sprintf("%s-instance-2", clusterName),
+					},
+					ExpectedInstancesToRemoveWithoutExclusion: []string{},
+					clusterName:  clusterName,
+					clusterLabel: fdbv1beta2.FDBClusterLabel,
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:138]: Is there a reason we add the cluster label here too but in other test cases we leave it empty?

File: kubectl-fdb/cmd/cordon_test.go [hunk 22]
```diff
@@ -21,121 +21,219 @@
...
 				testCase{
 					nodes:                     []string{"node-1", "node-2"},
+					WithExclusion:             false,
+					ExpectedInstancesToRemove: []string{},
+					ExpectedInstancesToRemoveWithoutExclusion: []string{
+						fmt.Sprintf("%s-instance-1", clusterName),
+						fmt.Sprintf("%s-instance-2", clusterName),
+					},
+					clusterName:  clusterName,
+					clusterLabel: fdbv1beta2.FDBClusterLabel,
```
Comment (johscheuer) [inline kubectl-fdb/cmd/cordon_test.go:150]: Same as above. We should either add it to all test cases or leave it empty and have a dedicated test case where we pass down the cluster name and the cluster label.

File: api/v1beta2/foundationdbcluster_types.go [hunk 23]
```diff
@@ -2275,3 +2275,18 @@ func (cluster *FoundationDBCluster) Validate() error {
 
 	return fmt.Errorf(strings.Join(validations, ", "))
 }
+
+// ContainsPod checks if the given Pod is part of the cluster or not.
+func (cluster *FoundationDBCluster) ContainsPod(pod corev1.Pod) bool {
```
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:2280]: Is that something we want to expose from the cluster struct of should this be rather part of the internal package? What do you think?

**Review by johscheuer — COMMENTED** [review]
Changes look good so far, a few comments.

File: kubectl-fdb/cmd/cordon.go (hunk 14 above)
Comment (09harsh) [reply kubectl-fdb/cmd/cordon.go:179]: mainly just to print all the errors at the end else user have to search the error with other logs. Another thing we can do here is print the error here and all the errors at the end as well.

File: kubectl-fdb/cmd/cordon.go (hunk 18 above)
Comment (09harsh) [reply kubectl-fdb/cmd/cordon.go]: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1496

**Review by johscheuer — APPROVED** [review]
Changes LGTM, could you please rebase the PR. That should make the e2e tests pass.

File: internal/pod_models.go [hunk 24]
```diff
@@ -1051,3 +1051,18 @@ func GetObjectMetadata(cluster *fdbv1beta2.FoundationDBCluster, base *metav1.Obj
 func GetPodDNSName(cluster *fdbv1beta2.FoundationDBCluster, podName string) string {
 	return fmt.Sprintf("%s.%s.%s.svc.%s", podName, cluster.Name, cluster.Namespace, cluster.GetDNSDomain())
 }
+
+// ContainsPod checks if the given Pod is part of the cluster or not.
+func ContainsPod(cluster *fdbv1beta2.FoundationDBCluster, pod corev1.Pod) bool {
```
[the remark this answers is not in the document]
Comment (09harsh) [reply internal/pod_models.go:1044]: done

**Review by johscheuer — APPROVED** [review]
Thanks, LGTM 👍
